File: algorithms/0343.integer_break/integer_break.cpp
class Solution {
public:
    int integerBreak(int n) {
        vector<int> dp(n+1, 0);
        dp[2] = 1;
        for (int i = 3; i <= n; ++i) {
            int cur_max = 0;
            for (int j = 1; j < i; ++j) {
                cur_max = max(cur_max, max(j * (i-j), j * dp[i-j]));
            }
            dp[i] = cur_max;
        }
        return dp[n];
    }
};

File: algorithms/0343.integer_break/integer_break.rs
impl Solution {
    pub fn integer_break(n: i32) -> i32 {
        let n = n as usize;
        let mut dp = vec![0; n + 1];
        dp[2] = 1;
        for i in 3..=n {
            for j in 1..i-1 {
                dp[i] = dp[i].max((i - j) * j).max(dp[i - j] * j);
            }
        }
        dp[n] as i32
    }
}